tactics/tactics.ml

```diff
@@ -4976,28 +4979,36 @@ let abstract_subproof id gk tac =
     Entries.(snd (Future.force const.const_entry_body)) in
   let solve =
     Proofview.tclEFFECTS effs <*>
-    exact_no_check (applist (lem, args))
+    tacK lem args
   in
   let tac = if not safe then Proofview.mark_as_unsafe <*> solve else solve in
   Sigma.Unsafe.of_pair (tac, evd)
   end }
 
+let abstract_subproof id gk tac ?(opaque=true) =
+  cache_term_by_tactic_then id gk ~opaque:opaque tac (fun lem args -> exact_no_check (applist (lem, args)))
+
 let anon_id = Id.of_string "anonymous"
 
-let tclABSTRACT name_op tac =
+let name_op_to_name name_op object_kind suffix =
   let open Proof_global in
-  let default_gk = (Global, false, Proof Theorem) in
-  let s, gk = match name_op with
+  let default_gk = (Global, false, object_kind) in
+  match name_op with
     | Some s ->
       (try let _, gk, _ = current_proof_statement () in s, gk
        with NoCurrentProof -> s, default_gk)
     | None   ->
       let name, gk =
 	try let name, gk, _ = current_proof_statement () in name, gk
 	with NoCurrentProof -> anon_id, default_gk in
-      add_suffix name "_subproof", gk
-  in
-  abstract_subproof s gk tac
+      add_suffix name suffix, gk
+
+let tclABSTRACT ?(opaque=true) name_op tac =
```

Comment:
It's not obvious to me that opacity should be default, it's just the first to be implemented. Since I tend to prefer explicitness that's enough of a reason for me to suggest making it a required argument.

Additional comments: 
- since `opaque` is passed as-is until `cache_term_by_tactic` if it's left optional that function should be the only one to have the default value (so if someone changes the `opaque=true` at `cache_term_by_tactic` they're not surprised when it does nothing since everything else still picks an unchanged default).

- `abstract_subproof` is generating a warning
`Warning 16: this optional argument cannot be erased.`
If `opaque` remains optional it should be moved towards the beginning of the arguments of this function. Maybe even if it becomes required, it's common in OCaml to have named arguments come first.